engine/access/state_stream/backend/backend_executiondata2_test.go

```diff
@@ -0,0 +1,1065 @@
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(nil, optimistic_sync.ErrBlockNotFound).
+			Once()
+
+		execDataRes, metadata, err :=
+			backend.GetExecutionDataByBlockID(ctx, block.ID(), s.criteria)
+		assert.Nil(s.T(), execDataRes)
+		assert.Nil(s.T(), metadata)
+		require.Error(s.T(), err)
+		require.True(s.T(), access.IsDataNotFoundError(err))
+	})
+
+	s.Run("execution result info returns fork abandoned", func() {
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(nil, optimistic_sync.ErrForkAbandoned).
+			Once()
+
+		execDataRes, metadata, err :=
+			backend.GetExecutionDataByBlockID(ctx, block.ID(), s.criteria)
+		assert.Nil(s.T(), execDataRes)
+		assert.Nil(s.T(), metadata)
+		require.Error(s.T(), err)
+		require.True(s.T(), access.IsPreconditionFailedError(err))
+	})
+
+	s.Run("execution result info returns unexpected error", func() {
+		// any unexpected error should be wrapped by access.RequireNoError, i.e., returned as-is
+		unexpected := assert.AnError
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(nil, unexpected).
+			Once()
+
+		exception := fmt.Errorf("failed to get execution result info for block %s: %w", block.ID(), unexpected)
+		ctxSignaler := irrecoverable.NewMockSignalerContextExpectError(s.T(), ctx, exception)
+		ctxIrr := irrecoverable.WithSignalerContext(ctx, ctxSignaler)
+
+		execDataRes, metadata, err :=
+			backend.GetExecutionDataByBlockID(ctxIrr, block.ID(), s.criteria)
+		assert.Nil(s.T(), execDataRes)
+		assert.Nil(s.T(), metadata)
+		assert.Error(s.T(), err)
+	})
+
+	s.Run("snapshot returns data not found", func() {
+		// first return a valid exec result info
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(func(blockID flow.Identifier, criteria optimistic_sync.Criteria) (*optimistic_sync.ExecutionResultInfo, error) {
+				info, err := s.executionResultProviderReal.ExecutionResultInfo(blockID, criteria)
+				return info, err
+			}).
+			Once()
+
+		// Snapshot not found maps to DataNotFoundError
+		s.executionStateCache.
+			On("Snapshot", mock.Anything).
+			Return(nil, storage.ErrNotFound).
+			Once()
+
+		execDataRes, metadata, err :=
+			backend.GetExecutionDataByBlockID(ctx, block.ID(), s.criteria)
+		assert.Nil(s.T(), execDataRes)
+		assert.Nil(s.T(), metadata)
+		require.Error(s.T(), err)
+		require.True(s.T(), access.IsDataNotFoundError(err))
+	})
+
+	s.Run("snapshot returns unexpected error", func() {
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(func(blockID flow.Identifier, criteria optimistic_sync.Criteria) (*optimistic_sync.ExecutionResultInfo, error) {
+				info, err := s.executionResultProviderReal.ExecutionResultInfo(blockID, criteria)
+				return info, err
+			}).
+			Once()
+
+		unexpected := assert.AnError
```

Comment:
not sure if this was in the original code, but I remember having several conversations about this with Andrii a while back. Let's avoid using `assert.AnError` when testing error handling. Instead, create a new specific error and check that you got it back.

Using overly generic data in tests can lead to false negatives hiding underlying failures. We need to be especially diligent when testing error conditions.

```suggestion
		exception := errors.New("unexpected error")
```